toolkit/tools/internal/userutils/userutils.go

```diff
@@ -0,0 +1,362 @@
+				err = file.Write(strings.Join(shadow, "\n"), ShadowFile)
+				return
+			}
+		}
+	}
+
+	return fmt.Errorf(`user "%s" not found when trying to change the password expiration date`, username)
+}
+
+func ConfigureUserGroupMembership(username string, primaryGroup string,
+	secondaryGroups []string, installChroot *safechroot.Chroot,
+) error {
+	var err error
+
+	// Update primary group
+	if primaryGroup != "" {
+		err = installChroot.UnsafeRun(func() error {
+			return shell.ExecuteLive(false /*squashErrors*/, "usermod", "-g", primaryGroup, username)
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Update secondary groups
+	if len(secondaryGroups) != 0 {
+		allGroups := strings.Join(secondaryGroups, ",")
+		err = installChroot.UnsafeRun(func() error {
+			return shell.ExecuteLive(false /*squashErrors*/, "usermod", "-a", "-G", allGroups, username)
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func ProvisionUserSSHCerts(username string, sshPubKeyPaths []string, installChroot *safechroot.Chroot) (err error) {
+	var (
+		pubKeyData []string
+		exists     bool
+	)
+	const squashErrors = false
+	const authorizedKeysTempFilePerms = 0644
+	const authorizedKeysTempFile = "/tmp/authorized_keys"
+	const sshDirectoryPermission = "0700"
+
+	// Skip user SSH directory generation when not provided with public keys
+	// Let SSH handle the creation of this folder on its first use
+	if len(sshPubKeyPaths) == 0 {
+		return
+	}
+
+	homeDir := UserHomeDirectory(username)
+	userSSHKeyDir := filepath.Join(homeDir, ".ssh")
+	authorizedKeysFile := filepath.Join(userSSHKeyDir, "authorized_keys")
+
+	exists, err = file.PathExists(authorizedKeysTempFile)
+	if err != nil {
+		logger.Log.Warnf("Error accessing %s file : %v", authorizedKeysTempFile, err)
+		return
+	}
+	if !exists {
+		logger.Log.Debugf("File %s does not exist. Creating file...", authorizedKeysTempFile)
+		err = file.Create(authorizedKeysTempFile, authorizedKeysTempFilePerms)
+		if err != nil {
+			logger.Log.Warnf("Failed to create %s file : %v", authorizedKeysTempFile, err)
+			return
+		}
+	} else {
+		err = os.Truncate(authorizedKeysTempFile, 0)
+		if err != nil {
+			logger.Log.Warnf("Failed to truncate %s file : %v", authorizedKeysTempFile, err)
+			return
+		}
+	}
+	defer os.Remove(authorizedKeysTempFile)
+
+	for _, pubKey := range sshPubKeyPaths {
+		logger.Log.Infof("Adding ssh key (%s) to user (%s)", filepath.Base(pubKey), username)
+		relativeDst := filepath.Join(userSSHKeyDir, filepath.Base(pubKey))
+
+		fileToCopy := safechroot.FileToCopy{
+			Src:  pubKey,
+			Dest: relativeDst,
+		}
+
+		err = installChroot.AddFiles(fileToCopy)
+		if err != nil {
+			return
+		}
+
+		logger.Log.Infof("Adding ssh key (%s) to user (%s) .ssh/authorized_users", filepath.Base(pubKey), username)
+		pubKeyData, err = file.ReadLines(pubKey)
+		if err != nil {
+			logger.Log.Warnf("Failed to read from SSHPubKey : %v", err)
+			return
+		}
+
+		// Append to the tmp/authorized_users file
```

Comment:
Should we check if the user is `root` then update `PermitRootLogin=yes`?

Reply:
The image builder toolkit doesn't do this.